### PR TITLE
New vrev update version scheme

### PR DESCRIPTION
**Pull request by adrianschroeter**
Still discussing this general solution for the vrev version race of build rpms when instantiating a new package update at the same time in next SP and in incident. This change avoids the race by enforcing the wanted scheme.

Other changes were depending on it, but actual change is the usage of extendrev parameter during
branching into an incident. (most code changes were due to rubocop complains)

This will get deployed when commitment from all stake holders is there.

This will lead to an enforced vrev scheme of

SP-M:GA   X.Y
SP-M:Update X.(Y+1).Z
SP-(M+1):GA X.(Y+2)

or real life

SP-2:GA 2.5
SP-2:Update 2.6.2
SP-3:GA 2.7

### CODE REVIEW

File: src/api/Gemfile.lock [hunk 1]
```diff
@@ -116,11 +116,11 @@ GEM
     faker (1.7.3)
       i18n (~> 0.5)
     feature (1.4.0)
+    ffi (1.9.17)
```
Comment (adrianschroeter) [inline src/api/Gemfile.lock]: That happened before, just the changes causing it in Gemfile.lock were patched different in Gemfile.lock. All what I did was a clean regenerate which should be reproducable. (But this is only cosmetic)

Comment (mdeniz) [inline src/api/Gemfile.lock]: If this is cosmetic... why did you added to the commit? I prefer to checkout the file in this case.

Comment (evanrolfe) [inline src/api/Gemfile.lock]: The same thing happens for me when I run bundle install, for some reason the ffi gem in Gemfile.lock is placed in the wrong line (alphabetically). This PR fixes that so IMO we should keep this change.

Comment (adrianschroeter) [inline src/api/Gemfile.lock]: @mdeniz it is an own commit ... did you mean we should put this into another pull request?

Fine for me for that one ....

Btw, buy-in and agreement exists not from all sides, but given that manual work on production data needs to happen also in the affected instance, we want to do it when I am back in 4 weeks.
So this will hang around some time, maybe a good idea to move the small Gemfile.lock in some other pull request then.

Comment (adrianschroeter) [inline src/api/Gemfile.lock]: I have put the three bugfixes in https://github.com/openSUSE/open-build-service/pull/3031 now

Comment (ChrisBr) [inline src/api/Gemfile.lock]: Yes please move it to another PR

**Comment by adrianschroeter** [comment]
JFYI, This will need to get applied in 2-3 weeks most likely. We just need to do it in time with other changes in our production setup.

**Comment by hennevogel** [comment]
Why don't you rebase?

**Comment by adrianschroeter** [comment]
it needs anyway the right timing and it is likely that i will need to rebase again until then ...

**Comment by hennevogel** [comment]
@adrianschroeter what are the conditions for the right timing? 😃

**Comment by adrianschroeter** [comment]
changed now in the project link vrevmode attribute way. Means it is not harmful to merge it anymore, we can enable it in future products in external obs instance later in a clean way...

**Comment by adrianschroeter** [comment]
I have no clue about the timeout errors in the rspec test suite, not sure if they are caused by my changes ...

**Comment by hennevogel** [comment]
@adrianschroeter your changes also have an influence on the `rspec` test suite. The timeouts are tests where the URL of the recorded backend answer doesn't match the requests that are fired while the tests run. As there is no backend running and answering you run into a timeout.

I would suggest you run the rspec test suite (which will produce new backend recordings) and adopt it. If you need help with this, let us know.

**Comment by bgeuken** [comment]
Could you document this new, additional revision mechanism? Be it in the [code](https://github.com/openSUSE/open-build-service/blob/master/src/api/app/helpers/maintenance_helper.rb), our [development wiki](https://github.com/openSUSE/open-build-service/wiki) or in our OBS documentaiton.

**Comment by bugfinder** [comment]
I've updated the patchset to apply again in https://github.com/bugfinder/open-build-service/tree/new_vrev_SP_scheme

**Comment by bugfinder** [comment]
I've updated the patchset to apply again in https://github.com/bugfinder/open-build-service/tree/new_vrev_SP_scheme

one new patch added in src/api/app/models/branch_package.rb to fix travis/lint

**Comment by hennevogel** [comment]
@adrianschroeter @bugfinder are you ever going to answer any of the questions from or [Björn](https://github.com/openSUSE/open-build-service/pull/3020#issuecomment-347798497) or [me](https://github.com/openSUSE/open-build-service/pull/3020#issuecomment-347234921) so we can get this merged?

**Comment by adrianschroeter** [comment]
@hennevogel I had not time yet to get the rspec test suite fully running. Problem is that recreating the cassettes show problems from commits not part of this pull request and that needs to be fixed as well.

We should maybe work together on that (but not today).

I don't see a question from Björn, just a request for writing documentation.

**Comment by bugfinder** [comment]
@hennevogel well, which question do you need answered from me ? at the moment I'm trying desparately to get enough enviromment up to get the testsuite running at all.
For the documentation side: all I know about it is: we have service pack layers, and this is an approach to make sure release numbers stay incremental between service packs while multiple ones (layers/service packs) are maintained in parallel.

the backend side was implemented in commit 1600cc643bda8291d6c314b7e15e7d37752cd62a

vrevmode=extend is an attribute for a project link and appends ".1" (as starter) to the vrev of a package and then the new last digit is increased on subsequent changes
vrevmode=unextend drops the last digit and increases the one before that.

**Comment by bugfinder** [comment]
@adrianschroeter I've fixed some cassettes updating the last patch in my branch again, testsuite is starting to look better https://travis-ci.org/openSUSE/open-build-service/jobs/329128273
https://github.com/openSUSE/open-build-service/pull/4346

**Comment by adrianschroeter** [comment]
@bugfinder merged and recreated one more

**Comment by adrianschroeter** [comment]
this breaks faster then we are able to fix it ... can we on a agree on a time to merge this so that we can fix the cassettes right before it?

(The rspec test suite is anyway not testing any functional change of this submissions, these are only side effects of code cleanup and incompabilities of changes already merged, but without recreated cassettes).

**Comment by adrianschroeter** [comment]
Marked as 2.9 critical or OBS 2.9 won't be able to use :Update projects of SLE 15 or Leap 15.

**Comment by hennevogel** [comment]
> We should maybe work together on that (but not today).

Okay so when? :-)

**Comment by adrianschroeter** [comment]
[Merge new vrev handling](https://trello.com/c/v1Dmk2KL/1316-merge-new-vrev-handling)

**Comment by bugfinder** [comment]
@adrianschroeter please pull from my branch again to update

**Comment by adrianschroeter** [comment]
done, and dropped also double and foreign commits ... too many people can push to my repo ...
